File: Cpp/Find Peak Element.cpp
// 8ms, 52.73%; 8.5MB, 95.45%
class Solution
{
public:
    int findPeakElement(const vector<int> &nums)
    {
        int size = nums.size();
        if (size == 1 || nums[0] > nums[1]) return 0;
        if (nums[size - 2] < nums[size - 1]) return size - 1;

        int left = 1, right = size - 1, mid;
        while (true) {
            mid = left + (right - left) / 2;
            if (nums[mid] > nums[mid - 1] && nums[mid] > nums[mid + 1]) break;
            if (nums[mid] < nums[mid - 1]) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return mid;
    }
};

File: Python3/Find Peak Element.py
# 48ms, 96.99%; 13.8MB, 5.88%
class Solution:
    def findPeakElement(self, nums: list) -> int:
        if len(nums) < 3: return nums.index(max(nums))
        if nums[0] > nums[1]: return 0
        if nums[-2] < nums[-1]: return len(nums) - 1

        left, right = 1, len(nums) - 1
        while True:
            mid = (left + right) // 2
            if nums[mid] > nums[mid - 1] and nums[mid] > nums[mid + 1]: return mid
            if nums[mid] < nums[mid - 1]:
                right = mid
            else:
                left = mid + 1
